### PR TITLE
scripts: Add build-blocks-manifest command

### PR DESCRIPTION
**Pull request by mreishus**
## What?

Adds a new `wp-scripts build-blocks-manifest` command.

## Why?

To help plugin authors take advantage of the new function `wp_register_block_metadata_collection()` offered in https://core.trac.wordpress.org/changeset/59132 . It requires all block.json files to be compiled into one php manifest file, which this command does.

## How?

Looks for block.json files, keys them off of the parent directory they belong to, then passes to `json2php` (new dependency alert) to write to a php file.

## Testing Instructions

* **In the Gutenberg repo**: Run `npm run build:blocks-manifest` and inspect the `build/blocks-manifest.php` file.
* **As wp-scripts**: I'm not sure, this is my first time working on this, but I hope it will 'just work' due to the PR following the conventions.

### CODE REVIEW

**Comment by mreishus** [comment]
Tag @felixarntz  ^ 
Is this what you had in mind? This is my first time working in wp-scripts.

File: packages/scripts/scripts/build-blocks-manifest.js
Comment (felixarntz) [inline packages/scripts/scripts/build-blocks-manifest.js:16]: Should the default be `build` or `src`? I can see why you chose `build`, but a counter argument would be that this means you always have to run `npm run build` _before_ you run `npm run build:blocks-manifest`.

To my knowledge, the `block.json` files are copied as is from `src` to `build`, so in that case I think `src` is a better default. WDYT?

**Review by felixarntz — COMMENTED** [review]
@mreishus Thanks for jumping on this to create the PR, it looks great to me so far. Kudos for already adding such extensive documentation! ❤️ 

I'm personally not the best candidate to review the JS command implementation, would be great to get reviews from people that were previously more involved in `@wordpress/scripts`. Maybe @gziolo you can help here?

File: packages/scripts/scripts/build-blocks-manifest.js
Comment (mreishus) [reply packages/scripts/scripts/build-blocks-manifest.js:16]: Hmm, I think I was influenced from my recent work in the WooCommerce plugin. That repo generates `block.json` files as a part of its main build process that are in directories named differently than their source values in `assets/`. Over there, the only way I could generate the right information was to pull from `build/` while also writing to `build/`. That may be atypical though.

Comment (mreishus) [reply packages/scripts/scripts/build-blocks-manifest.js:16]: I was about to move it to `src` here, but I found that in order to keep the `npm run build:blocks-manifest` command working in the root, I'd have to look through 4 src directories:
```
./packages/blocks/src/
./packages/widgets/src/
./packages/block-library/src/
./packages/edit-widgets/src/
```
We don't have a way to specify multiple input dirs. It may be the `build->build` after is a useful pattern for monorepos?

Comment (gziolo) [reply packages/scripts/scripts/build-blocks-manifest.js:16]: This one is interesting. When you use `wp-scripts build` then `block.json` files get copied from `src` to build and sometimes they even can get modified:

https://github.com/WordPress/gutenberg/blob/79940e0ca94cd814631e9a29f2d9178352246c70/packages/scripts/config/webpack.config.js#L314-L345

In effect, I think it's a valid approach to source the `block.json` from the `build` folder to put them in the `build` folder. In the long run, these probably should be an option on the `wp-scripts build --use-manifest`. However, we can get there in steps.

File: package.json
Comment (gziolo) [inline package.json]: This could go next in another PR to make it all simpler to land 😄

**Review by sirreal — COMMENTED** [review]
The code looks reasonable to me. It would be good to add a CHANGELOG entry.

A basic test would be nice, it could be a snapshot of the expected result. DEWP has some testing like this that could serve as inspiration:

https://github.com/WordPress/gutenberg/blob/trunk/packages/dependency-extraction-webpack-plugin/test/build.js

**Review by sirreal — APPROVED** [review]
I tested this out in Gutenberg and it seems to work well. I did list a couple nice-to-haves in [my previous review](https://github.com/WordPress/gutenberg/pull/65866#pullrequestreview-2357318949) but nothing blocking.

File: packages/scripts/scripts/build-blocks-manifest.js
Comment (felixarntz) [reply packages/scripts/scripts/build-blocks-manifest.js:16]: That all makes sense, let's keep using `build` for now.

**Review by felixarntz — APPROVED** [review]
Giving this a 2nd approval, pending the minor feedback. Thanks @mreishus!

File: package.json
Comment (gziolo) [reply package.json]: This could be removed for now as it isn’t connected anywhere.

**Review by gziolo — APPROVED** [review]

File: package.json
Comment (mreishus) [reply package.json]: Thank you, removed here: e24747ee85

**Comment by mreishus** [comment]
* added snapshot test 389b12858de3390d88fcad5f177ec0e590ffeca4 
* rebased and solved lockfile conflict (by dropping the npm install commit and making a new one)
* added changelog 840742a66fa3b125b6be3bc82e1a018297604a23

File: packages/scripts/scripts/build-blocks-manifest.js
Comment (jsnajdr) [inline packages/scripts/scripts/build-blocks-manifest.js:36]: The script should first check if the `build` directory exists and abort with error if it doesn't.

It makes sense to run this script only if the blocks were already built and the `build` directory was created. If you forgot to run the build first, you should be told about it, instead of silently creating an empty manifest file.

Comment (mreishus) [reply packages/scripts/scripts/build-blocks-manifest.js:36]: :+1: Check https://github.com/WordPress/gutenberg/pull/65866/commits/e673f314c2f0cd3db49471c2892257889d4bff56

**Comment by felixarntz** [comment]
@jsnajdr Could you please take another look at this? Ideally, we could merge this before Thursday, as I am currently working on a dev note post for the change for WordPress 6.7, and it would make sense to have this available to reference for developers already in that post.

See https://make.wordpress.org/core/?p=115864&preview=1&_ppp=848634d99d for the public preview of that dev note. @gziolo @mreishus Could you please also review my draft and let me know your feedback?

**Comment by gziolo** [comment]
> Ideally, we could merge this before Thursday, as I am currently working on a dev note post for the change for WordPress 6.7, and it would make sense to have this available to reference for developers already in that post.

Merging this PR is not enough to publish it to npm. The next package release will happen next week on Wednesday. Let me know if you would like to publish it separately on demand earlier.

> Could you please also review my draft and let me know your feedback?

I'm checking the draft next.

**Comment by gziolo** [comment]
> See https://make.wordpress.org/core/?p=115864&preview=1&_ppp=848634d99d for the public preview of that dev note.

I'm not sure what is the best place for feedback. I can start here and move it elsewhere upon request:
- It needs to be clarified `wp_register_block_metadata_collection()` is an additional step on top of the existing registration process. In effect, I would include the snippet illustrating how the blocks get registered using:
    ```php
    wp_register_block_metadata_collection(
        WP_PLUGIN_DIR . '/my-block-library/build',
        WP_PLUGIN_DIR . '/my-block-library/build/blocks-manifest.php'
    );
    register_block_type( WP_PLUGIN_DIR . '/my-block-library/build/pricing' );
    register_block_type( WP_PLUGIN_DIR . '/my-block-library/build/services', array( 'render_callback' => 'my_block_library_render_callback' ) ); // It should still be possible to pass additional settings, so might be worth highlighting.
    register_block_type( WP_PLUGIN_DIR . '/my-block-library/build/testimonial/block.json' ); // Does it work with full path, too?
   ```
- As noted in the code snippet, some notes about the path defined to the `block.json` file when using the collection would be helpful.
- Totally optional, illustrate that passing additional settings when registering the block is still compatible with the metadata collection.

**Comment by mreishus** [comment]
```php
register_block_type( WP_PLUGIN_DIR . '/my-block-library/build/testimonial/block.json' ); // Does it work with full path, too?
```
Indeed, it works with full paths as well.

For the draft, it looks good! I agree with the "clarify it's an additional step" from @gziolo. For myself, I have no suggestions, other than optionally, you may want to also include https://make.wordpress.org/core/2022/10/07/improved-php-performance-for-core-blocks-registration/ when showing previous context.

Also, I'm sure you guys already know this, but just to be un-ambiguous, I don't have the ability to merge this PR. :)

**Comment by gziolo** [comment]
Everything works as intended:

Screenshot 2024-10-17 at 13 13 51

**Comment by gziolo** [comment]
Once PR lands in `trunk`, I will cherry pick to `wp/latest` and trigger npm publishing.

**Review by jsnajdr — APPROVED** [review]
This looks good to me, thanks @mreishus for implementing the suggestion to abort right away when the `build` directory doesn't exist.

**Comment by gziolo** [comment]
Publishing to npm with https://github.com/WordPress/gutenberg/actions/runs/11385894200.

**Pull request merged**
